heat/core/statistics.py

```diff
@@ -626,77 +627,94 @@ def maximum(x1, x2, out=None):
+                    )
+                )
+        elif splits.count(None) == 1:
+            # one DNDarray non-distributed, OK only for split along broadcastable axis
+            split_none = splits.index(None)
+            split_not_none = abs(split_none - 1)
+            if in_x[split_none].size > 1 and shapes[split_none][splits[split_not_none]] != 1:
+                raise ValueError(
+                    "input DNDarrays must be distributed along the same axis, currently x1.split is {}, x2.split is {}".format(
+                        splits[0], splits[1]
+                    )
+                )
+            else:
+                split = splits[split_not_none]
+                comm = comms[split_not_none]
     else:
         split = None
+        comm = comms[0]
 
-    # locally: apply torch.max(x1, x2)
-    output_lshape = stride_tricks.broadcast_shape(x1.lshape, x2.lshape)
-    lresult = factories.empty(output_lshape, dtype=x1.dtype)
-    lresult._DNDarray__array = torch.max(x1._DNDarray__array, x2._DNDarray__array)
-    lresult._DNDarray__dtype = types.promote_types(x1.dtype, x2.dtype)
-    lresult._DNDarray__split = split
-    if x1.split is not None or x2.split is not None:
-        if x1.comm.is_distributed():  # assuming x1.comm = x2.comm
-            output_gshape = stride_tricks.broadcast_shape(x1.gshape, x2.gshape)
-            result = factories.empty(output_gshape, dtype=x1.dtype)
-            x1.comm.Allgather(lresult, result)
-            # TODO: adopt Allgatherv() as soon as it is fixed, Issue #233
-            result._DNDarray__dtype = lresult._DNDarray__dtype
-            result._DNDarray__split = split
-
-            if out is not None:
-                if out.shape != output_gshape:
-                    raise ValueError(
-                        "Expecting output buffer of shape {}, got {}".format(
-                            output_gshape, out.shape
-                        )
-                    )
-                out._DNDarray__array = result._DNDarray__array
-                out._DNDarray__dtype = result._DNDarray__dtype
-                out._DNDarray__split = split
-                out._DNDarray__device = x1.device
-                out._DNDarray__comm = x1.comm
-
-                return out
-            return result
-
+    # sanitize out
     if out is not None:
-        if out.shape != output_lshape:
+        if not isinstance(out, dndarray.DNDarray):
+            raise TypeError("expected `out` to be None or a DNDarray, but was {}".format(type(out)))
+        if out.gshape != output_gshape:
             raise ValueError(
-                "Expecting output buffer of shape {}, got {}".format(output_lshape, out.shape)
+                "Expecting output buffer of shape {}, got {}".format(output_gshape, out.shape)
+            )
+        if out.split is not split:
+            raise ValueError(
+                "Split axis of output buffer is inconsistent with split semantics (see documentation)."
             )
```

Comment:
would it be possible to make this a `sanitize_out()` function? these seems like the kind of logic that we call could use for the out parameter in most function. maybe it could be combined with the setting logic as well? (lines 711-716)

Reply:
Yes, I think I should stop procrastinating on that sanitation module (#468)

Reply:
sanitation module with basic functions (including sanitize_out) is in now